samples: nrf9160: gnss: rename sample and add nRF Cloud A-GPS/P-GPS support

tokangas opened the pull request:
Renamed GPS sample to GNSS sample.

Added support for nRF Cloud A-GPS and P-GPS using REST API.

When assistance is used, LTE now remains connected all the time by default. Both eDRX and PSM are requested from the network. It's possible to enable the old behavior using a new Kconfig option so that LTE is only activated when needed to fetch A-GPS data.

LTE link control is used to configure modem also when assistance is not used.

peknis commented on doc/nrf/libraries/others/supl_os_client.rst:
```diff
@@ -120,8 +120,8 @@ API callback functions
-You can find an implementation example in the :ref:`gps_with_supl_support_sample` sample.
-The implementation is located in the :file:`samples\\nrf9160\\gps\\src\\` files.
+You can find an implementation example in the :ref:`gnss_sample` sample.
+The implementation is located in the :file:`samples\\nrf9160\\gnss\\src\\` files.
```
```suggestion
The implementation is located in the :file:`samples\nrf9160\gnss\src\` directory.
```

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -161,6 +161,20 @@ nRF9160 samples
 
+* :ref:`gnss_sample` sample:
+
+  * Renamed from GPS sample.
```
```suggestion
  * Renamed. The previous name was nRF9160: GPS with SUPL client library.
```

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -161,6 +161,20 @@ nRF9160 samples
+
+  * Renamed from GPS sample.
+  * Added support for nRF Cloud A-GPS and P-GPS.
+  * LTE now remains connected to the network all the time when assistance is enabled. It's possible to enable old behavior with A-GPS using a Kconfig option.
```
```suggestion
  * LTE now remains connected to the network all the time when assistance is enabled. You can enable the old behavior with A-GPS using a Kconfig option.
```

peknis commented on samples/nrf9160/gnss/Kconfig:
```diff
@@ -0,0 +1,94 @@
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_NRF9160DK_NRF9160_NS || BOARD_THINGY91_NRF9160_NS) && GNSS_SAMPLE_ANTENNA_ONBOARD
+	default "AT\%XCOEX0" if (BOARD_NRF9160DK_NRF9160_NS || BOARD_THINGY91_NRF9160_NS) && GNSS_SAMPLE_ANTENNA_EXTERNAL
+	help
+	  Defines what is the AT%XCOEX0 command to be sent to GNSS module. Leave
```
```suggestion
	  Defines the AT%XCOEX0 command to be sent to GNSS module. Leave
```

peknis commented on samples/nrf9160/gnss/Kconfig:
```diff
@@ -0,0 +1,94 @@
+	default "AT\%XMAGPIO=1,0,0,1,1,1574,1577" if BOARD_NRF9160DK_NRF9160_NS
+	default "AT\%XMAGPIO=1,1,1,7,1,746,803,2,698,748,2,1710,2200,3,824,894,4,880,960,5,791,849,7,1565,1586" if BOARD_THINGY91_NRF9160_NS
+	help
+	  Defines what is the AT%XMAGPIO command to be sent to GNSS module. Leave
```
```suggestion
	  Defines the AT%XMAGPIO command to be sent to GNSS module. Leave
```

peknis commented on doc/nrf/libraries/others/supl_os_client.rst:
```diff
@@ -120,8 +120,8 @@ API callback functions
 If your application uses the SUPL client library, you must implement these callback functions.
-You can find an implementation example in the :ref:`gps_with_supl_support_sample` sample.
-The implementation is located in the :file:`samples\\nrf9160\\gps\\src\\` files.
+You can find an implementation example in the :ref:`gnss_sample` sample.
```
```suggestion
For an example of the implementation, see the :ref:`gnss_sample` sample.
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -1,23 +1,24 @@
-The GPS sample demonstrates how to retrieve `GNSS`_ data.
-If Secure User-Plane Location (SUPL) support is enabled, it also shows how to improve the fix accuracy and fix speed with `A-GPS`_ data from a SUPL server.
-See :ref:`supl_client` for information on enabling SUPL support for the sample.
+The GNSS sample demonstrates how to retrieve `GNSS`_ data.
```
```suggestion
This sample demonstrates how to retrieve `GNSS`_ data.
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -1,23 +1,24 @@
-If Secure User-Plane Location (SUPL) support is enabled, it also shows how to improve the fix accuracy and fix speed with `A-GPS`_ data from a SUPL server.
-See :ref:`supl_client` for information on enabling SUPL support for the sample.
+The GNSS sample demonstrates how to retrieve `GNSS`_ data.
+It also shows how to improve fix speed and accuracy with :ref:`lib_nrf_cloud_agps` and how to use :ref:`lib_nrf_cloud_pgps`.
```
```suggestion
It also shows how to improve fix speed and accuracy with the :ref:`lib_nrf_cloud_agps` library and how to use the :ref:`lib_nrf_cloud_pgps` library.
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -31,8 +32,8 @@ In the default mode, the sample displays information from both PVT (Position, Ve
 
-SUPL support can be enabled for both the default mode (PVT and NMEA) and the NMEA-only mode.
-When the SUPL support is enabled, the sample receives an A-GPS data request notification from the GNSS module, and it starts downloading the A-GPS data requested by the GNSS module.
+A-GPS and P-GPS support can be enabled for both the default mode (PVT and NMEA) and the NMEA-only mode.
```
```suggestion
You can enable A-GPS and P-GPS support for both the default mode (PVT and NMEA) and the NMEA-only mode.
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -142,28 +139,28 @@ After programming the sample and all the prerequisites to the development kit, y
-      i. The sample switches to LTE and connects to a SUPL server.
-      #. The A-GPS data gets downloaded.
-      #. The sample continues after the SUPL session is complete.
+      i. The sample downloads the requested assistance data if needed (with P-GPS the data may already be available in flash memory).
```
```suggestion
      i. The sample downloads the requested assistance data if needed (with P-GPS, the data may already be available in the flash memory).
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -142,28 +139,28 @@ After programming the sample and all the prerequisites to the development kit, y
-      #. The A-GPS data gets downloaded.
-      #. The sample continues after the SUPL session is complete.
+      i. The sample downloads the requested assistance data if needed (with P-GPS the data may already be available in flash memory).
+      #. The sample continues after the download has been completed.
```
```suggestion
      #. The sample continues after the download has completed.
```

peknis commented on samples/nrf9160/gnss/README.rst:
```diff
@@ -1,23 +1,24 @@
+The GNSS sample demonstrates how to retrieve `GNSS`_ data.
+It also shows how to improve fix speed and accuracy with :ref:`lib_nrf_cloud_agps` and how to use :ref:`lib_nrf_cloud_pgps`.
+Assistance data is downloaded from nRF Cloud using `nRF Cloud's REST-based device API <nRF Cloud REST API_>`_.
+The sample also supports using SUPL client library for downloading A-GPS data from a SUPL server.
```
```suggestion
The sample also supports using the SUPL client library for downloading A-GPS data from a SUPL server.
```

peknis commented on samples/nrf9160/gnss/src/assistance.h:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @brief Initializes the assistance module.
+ *
+ * @param[in] assistance_work_q Work queue which can be used by the assistance module.
```
```suggestion
 * @param[in] assistance_work_q Work queue that can be used by the assistance module.
```

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6010
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

plskeggs commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,345 @@
+agps_exit:
+	assistance_active = false;
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+
```
I believe you need to call `nrf_cloud_rest_disconnect(&rest_ctx)`

tokangas replied:
`rest_ctx.keep_alive` is set to `false`, so the connection is automatically disconnected by the nRF Cloud REST lib.

plskeggs commented on samples/nrf9160/gnss/src/main.c:
```diff
@@ -346,28 +348,32 @@ static void print_fix_data(struct nrf_modem_gnss_pvt_data_frame *pvt_data)
 					       pvt_data->datetime.seconds);
```
In my PR you had suggested including speed accuracy, and using %04u for the year.

tokangas replied:
I had already started modifying the sample output, but then decided that it's better to do a separate PR for that. I'm going to add some more information from the PVT at the same time, so let's keep that separate.

tokangas commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,345 @@
+
+	assistance_active = true;
+
+	validity_time = 5 * 60; /* 5 minutes */
```
PR #5974 changes the default value to 10 minutes, so I'll switch to using the default value.

tokangas commented on samples/nrf9160/gnss/Kconfig:
```diff
@@ -0,0 +1,94 @@
+	depends on !NRF_CLOUD_PGPS
+	help
+	  Activates LTE only when it is needed to fetch A-GPS data. This is not supported when
+	  P-GPS is enabled.
```
P-GPS works in a different way and things get more complicated. With A-GPS it's enough to activate LTE before requesting A-GPS data and deactivate LTE after that function returns. However, with P-GPS the prediction download happens "in the background" and a different mechanism would be needed to determine when LTE needs to be activated and deactivated. With both A-GPS and P-GPS enabled it gets even more complicated.

I considered implementing support for this but didn't see it worth the effort. The way the GPS sample has been implemented has been a bit awkward to begin with. The preferred way is to keep LTE connected with PSM and/or eDRX enabled. After all, if the application is acquiring GNSS fixes, it most likely also sends those to the cloud, for which it needs the network connection. I even considered not implementing support for "LTE on-demand" at all, but decided to keep that with A-GPS because that was straight forward to implement.

peknis commented on doc/nrf/libraries/others/supl_os_client.rst:
```diff
@@ -120,8 +120,8 @@ API callback functions
-You can find an implementation example in the :ref:`gps_with_supl_support_sample` sample.
-The implementation is located in the :file:`samples\\nrf9160\\gps\\src\\` files.
+For an example of the implementation, see the :ref:`gnss_sample` sample.
+The implementation is located in the :file:`samples\nrf9160\gnss\src\` directory.
```
```suggestion
The implementation is located in the :file:`samples/nrf9160/gnss/src/` directory.
```
Maybe we need slashes instead of backslashes. Looks like we have them that way in other docs. Not sure this caused the build fail, but could well be this.

hiltunent commented on samples/nrf9160/gnss/src/main.c:
```diff
@@ -231,45 +109,160 @@ static void gnss_event_handler(int event)
-		printk("Failed to initialize modem\n");
-		return -1;
-	}
+	return assistance_init(&agps_work_q);
```
You now have assistance_init() and assist_init() function names. Could you call one of them e.g. pgps_init() or similar as it seems to be mostly about pgps?

tokangas replied:
`assistance.h` is a generic interface that is used with both nRF Cloud and SUPL, so I don't want to change that. I tried to think about a better name for the initialization function in `main.c`, but couldn't come up with any...

What if I change `assist_init()` to `sample_init()`? Although related to assistance, it does initialize the work queue and work item. And actually `sample_init()` could even call `modem_init()` and `gnss_init_and_start()` instead of calling those directly from main.

hiltunent replied:
I see...work_init() could be one candidate as well. Well, you pick whatever you think is most describing. No need to change it just for sake of doing something.

hiltunent commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,340 @@
+		return err;
+	}
+
+	char resp_buf_cellid[MODEM_INFO_MAX_RESPONSE_SIZE];
+	char resp_buf_tac[MODEM_INFO_MAX_RESPONSE_SIZE];
+	char resp_buf_cops[MODEM_INFO_MAX_RESPONSE_SIZE];
```
Same comment here that popped up with Location lib. You can save mem by using only one buffer and parsing the data as you go.

tokangas replied:
Will do.

tokangas commented on samples/nrf9160/gnss/src/main.c:
```diff
@@ -231,45 +109,160 @@ static void gnss_event_handler(int event)
+		return;
+	}
+
+	printk("LTE activated\n");
```
"LTE activated" doesn't mean that modem is yet connected, it just indicates that LTE was enabled. The LTE LC event handler prints "Connected to LTE network" when connection has been established. I agree that the term "activated" can be confusing, but the same terminology is used in LTE LC and the modem AT specification.

hiltunent replied:
Okay, no worries then.

tokangas commented on doc/nrf/ug_nrf91_features.rst:
```diff
@@ -346,10 +346,10 @@ Samples using GNSS in |NCS|
 * :ref:`serial_lte_modem` application - Uses AT commands to start and stop GNSS and has support for nRF Cloud A-GPS and P-GPS. The application displays tracking and GNSS fix information in the serial console.
 * :ref:`agps_sample` sample - Uses nRF Cloud for A-GPS by default and can be configured to use SUPL. The sample obtains GNSS fixes and transmits them to nRF Cloud.
-* :ref:`gps_with_supl_support_sample` sample - Does not use A-GPS by default but can be configured to use SUPL. The sample displays tracking, fix information, and NMEA strings in the serial console.
+* :ref:`gnss_sample` sample - Does not use assistance by default but can be configured to use nRF Cloud A-GPS or P-GPS or a combination of both. The sample displays tracking, fix information, and NMEA strings in the serial console.
```
@peknis, is this change ok?

tokangas replied:
👍 I'll update it.

hiltunent commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,340 @@
+		break;
+	}
+}
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
```
Same define ends and begins again, remove these lines

tokangas replied:
True. I'll also move the semaphore definition to the beginning of the file and make it static.

tokangas commented on samples/nrf9160/gnss/src/main.c:
```diff
@@ -4,50 +4,33 @@
 static struct k_work_q agps_work_q;
-static struct k_work get_agps_data_work;
+static struct k_work agps_data_get_work;
+static bool requesting_assistance;
```
Probably a good idea, I'll make it volatile.

tokangas commented on samples/nrf9160/gnss/src/main.c:
```diff
@@ -231,45 +109,160 @@ static void gnss_event_handler(int event)
+	err = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_DEACTIVATE_LTE);
+	if (err) {
+		printk("Failed to deactivate LTE, error: %d\n", err);
+
```
I'll remove empty lines before `return` from both `lte_activate()` and `lte_deactivate()`.

jtguggedal commented on samples/nrf9160/gnss/prj.conf:
```diff
@@ -12,11 +13,21 @@ CONFIG_NEWLIB_LIBC=y
+#CONFIG_GNSS_SAMPLE_NMEA_ONLY=y
+#CONFIG_GNSS_SAMPLE_ANTENNA_EXTERNAL=y
+# Enable to use nRF Cloud A-GPS
+#CONFIG_GNSS_SAMPLE_ASSISTANCE_NRF_CLOUD=y
```
I think we could leave out the out-commented configs and rather present them in the sample docs

jtguggedal commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,338 @@
+
+	int err;
+
+	printk("Sending request for P-GPS predictions to nRF Cloud...\n");
```
The logger could be used here, though that may require the whole sample to be converted

tokangas replied:
I didn't want to switch to logger, because I think the output is clearer without (if you consider the GNSS data which is printed once a second). The output related to A-GPS and P-GPS could use logger. Do you think I should change that?

jtguggedal replied:
Hm, it could be beneficial in the case where NMEA printing is enabled, perhaps. Then the assistance printing can be turned off with the logger to avoid cluttering the output

tokangas replied:
That's a good point. Output related to assistance is tricky, because it interferes with the GNSS PVT/NMEA output. The current way of handling this is that PVT/NMEA output is paused while assistance data is being downloaded. Logger would make it possible to suppress the assistance related output when needed, but this would need to be somehow taken into account in the PVT/NMEA output so that it isn't unnecessarily paused when assistance output is disabled. I think this could be done by checking the logging level. I'll look into this.

plskeggs replied:
My version of this sample did switch to logger, btw...  Though I didn't test using an external NMEA visualizer.  What program do you use for that @tokangas ?

jtguggedal commented on samples/nrf9160/gnss/src/assistance.c:
```diff
@@ -0,0 +1,338 @@
+		.auth = jwt_buf,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0, /* use CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE */
```
Will it default to using `CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE` when it's set to zero?
If so, the comment could be something like `/* Defaults to CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE when 0 */`

tokangas commented:
@plskeggs I will try using logger in a separate PR, because I've anyway plans to at least add more information from the PVT to the output. I've been using [VisualGPSView](https://www.visualgps.net/) for visualizing NMEA output. If logger interferes with the NMEA parsing, we could use `printk()` in the NMEA only mode and otherwise logger.

tokangas commented:
I've added the configuration part to `README.rst` and updated `prj.conf` a bit.

@jtguggedal The `prj.conf` now has two options which are not commented out, but set to `n`. I think it would be good to have these two visible, because these are the two most important options for the sample. If this is not ok I can of course remove those, but I'll give it a try. :)

jtguggedal commented:
> @jtguggedal The `prj.conf` now has two options which are not commented out, but set to `n`. I think it would be good to have these two visible, because these are the two most important options for the sample. If this is not ok I can of course remove those, but I'll give it a try. :)

I'm good with keeping those options as disabled in prj.conf

rlubos reviewed (approved):
LGTM
